### ISSUE TITLE
Unexpected Behavior with Jittable class

### ISSUE THREAD

**Issue by schlagercollin** [deepmind/distrax#162]
Hello,

First off, thanks for your hard work in developing this package. It's great to have a JAX-native implementation of TFP. 

In working with the library, I noticed some unexpected behavior using the `Jittable` class. The underlying problem is that the `Jittable` class registers all of the instance's data under the `aux_data` of the Jax PyTree (as opposed to the `children`) [Line 35](https://github.com/deepmind/distrax/blob/f2ef70c39350c7487b109176340ec506d8406e00/distrax/_src/utils/jittable.py#L35). My understanding is that the `aux_data` is meant to store object metadata, while the actual array-like data should be stored in the `children` of the PyTree node (see [JAX docs](https://jax.readthedocs.io/en/latest/pytrees.html)).

As a result of the current implementation of `Jittable`, a few undesirable things happen.

### 1) Jittable Functions Can't Be Re-used Across Objects of the Same Shape 

The benefit of jitting is that compiled code can be re-used on similar inputs. However, the current implementation of `Jittable` does not allow this. Because data (like a model's parameters) are stored in the `aux_data` (metadata) field of the Jax PyTree, any changes to those parameters will trigger a re-compile since JAX thinks you're passing in a different type of object.  

In fact, we don't even get that far using `Jittable`, since the code fails when attempting to compare the metadata (to determine if it's different --> thus needing a re-compile). Here's a quick example: I create two Bernoulli model objects with the same shaped parameters. I then create a simple jitted function (to sample from the provided model) and run it on `bern_1` followed by `bern_2`. 

```python
import distrax  # __version__ = 0.1.2

import jax.numpy as np
import jax.random as jr
from jax import jit

bern_1 = distrax.Bernoulli(logits=np.ones((3,)))
bern_2 = distrax.Bernoulli(logits=np.zeros((3,)))

@jit
def jitted_function(model, seed):
    return model.sample(seed=seed, sample_shape=(5,))

seed = jr.PRNGKey(0)
jitted_function(bern_1, seed)  # jitted_function is compiled here

jitted_function(bern_2, seed)  # throws ValueError 
```

The ValueError thrown by the last line:
```python
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

### 2) Modifying Object Data Leads to Escaped Tracers

Another issue with registering parameter-like data in the `aux_data` field of a PyTree is that modifying values inside a jitted function leads to undesirable behavior. 

In this minimal example, I make use of the `DummyJittable` object from the jittable test harness. I define a simple jitted function that adds one to the object's parameters. This function can be seen as a basic analogy for a function that increments a model's parameters in some optimization routine like SGD, e.g.

The code runs, but printing the object's parameters after the function shows that Jax Tracers have escaped the jitted function's context. 

```python
from distrax._src.utils.jittable import Jittable

class DummyJittable(Jittable):
  """
  Taken from distrax._src.utils.jittable_test
  """
  def __init__(self, params):
    self.name = 'dummy'  # Non-JAX property, cannot be traced.
    self.data = {'params': params}  # Tree property, must be traced recursively.

dummy = DummyJittable(np.zeros((5,)))

@jit
def add_one_to_params(obj):
    print("jitting...")
    obj.data["params"] = obj.data["params"] + 1 
    return obj

dummy_out = add_one(dummy)
print(dummy_out.data)
```

```python
>>> {'params': Traced<ShapedArray(float32[5])>with<DynamicJaxprTrace(level=0/1)>}
```

If we wanted to modify these parameters outside a jitted function's context, we would get an error.

```python
dummy_out.data["params"] - 1 # throws UnexpectedTracerError
```

```python
>>> UnexpectedTracerError: Encountered an unexpected tracer. A function transformed by JAX had a side effect, allowing for a reference to an intermediate value with shape (5,) and dtype float32 to escape.
JAX transformations require that functions explicitly return their outputs, and disallow saving intermediate values to global state.
The function being traced when the value leaked was add_one at /var/folders/kc/slznj52d4p57wn5mdxm530080000gn/T/ipykernel_78765/2780791423.py:1 traced for jit.
```

### How do we fix this behavior?

Non-metadata for a Jittable object should be registered in the `children` field of its PyTree. This solves the two problems presented above.

Here's a super-simple example of re-writing the `DummyJittable` to behave as expected.

```python

from jax.tree_util import register_pytree_node_class

@register_pytree_node_class
class AlternativeDummyJittable:
    def __init__(self, params):
        self.data = {'params': params}
        
    def tree_flatten(self):
        params = self.data["params"]
        return (params,), None  # (children, aux_data) <-- notice params goes in children
    
    @classmethod
    def tree_unflatten(cls, aux_data, children):
        params, = children
        return cls(params)
```

With this definition, the jitted function behaves as expected.

```python
@jit
def add_one(obj):
    print("jitting...")
    obj.data["params"] = obj.data["params"] + 1 
    return obj

alt = AlternativeDummyJittable(np.zeros((5,)))
alt_out = add_one(alt)  # function compiles here
print(alt_out.data)  # prints JAX array as opposed to tracer values

# we can re-use the compiled function on an object with different parameters / same shape!
alt_different_params = AlternativeDummyJittable(np.ones((5,)))
alt_out = add_one(alt_different_params)  # no need to re-compile
print(alt_out.data)  # prints JAX array as opposed to tracer values
```

Hopefully the above makes sense. I'm happy to elaborate on any points made above. Currently, the `Jittable` class is misleading and breaks on standard logic of a "jittable" object. Object data like parameters should be registered in the `children` field of a PyTree instead of the `aux_data` field.

Thank you!
-Collin

**Comment by bayerj** [comment]
I got bitten by this today. LeakedTracer errors are the worst, and really hard to pin down and fix.

Actually, I feel that the [flax folks got it right](https://github.com/google/flax/blob/main/flax/struct.py) with `flax.struct.dataclass` and `flax.struct.field`. It is very easy to handle, since whether an attribute is part of the pytree or meta is made explicit.

**Comment by ojwenzel** [comment]
Your work on this native JAX implementation for TFP is also much appreciated on my end as well! Thanks a lot. 

Unfortunately, I hit the same error today, and it does keep me from going full distrax. A solution would be very helpful!

**Comment by jakebruce** [comment]
Thank you @schlagercollin for an excellent and thorough report. We believe the issue is now fixed in #168 and have incorporated elements of your report as new test cases.

**Comment by jakebruce** [comment]
Thanks @schlagercollin for your very thorough report. We've incorporated your examples as test cases and pushed a change (#177) that we believe addresses the issue.
